api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java

```diff
@@ -425,85 +431,177 @@ public PatientIdentifierType getPatientIdentifierType(Integer patientIdentifierT
+				if (patientIds.size() >= 1) {
+					Query query = sessionFactory.getCurrentSession().createQuery(
+							"from Patient p1 where p1.patientId in (:ids)");
+					query.setParameterList("ids", patientIds);
+					patients = query.list();
+				}
 			}
-			
-			Class personName = PersonName.class;
-			Set<String> personNameFieldNames = new HashSet<String>(personName.getDeclaredFields().length);
-			for (Field f : personName.getDeclaredFields()) {
-				personNameFieldNames.add(f.getName());
-				log.debug(f.getName());
+			catch (HibernateException ex) {
+				log.warn("Hibernate exception: " + ex.getStackTrace().toString());
+				throw ex;
+
 			}
-			
-			Class identifier = PatientIdentifier.class;
-			Set<String> identifierFieldNames = new HashSet<String>(identifier.getDeclaredFields().length);
-			for (Field f : identifier.getDeclaredFields()) {
-				identifierFieldNames.add(f.getName());
-				log.debug(f.getName());
+			finally {}
+		}
+
+		sortDuplicatePatients(patients, patientIds);
+		return patients;
+	}
+
+	private String getDuplicatePatientsSQLString(List<String> attributes) {
+		String outerSelect = "select distinct t1.patient_id from patient t1 ";
+
+		Class patient = Patient.class;
+		Set<String> patientFieldNames = new HashSet<String>(patient.getDeclaredFields().length);
+		for (Field f : patient.getDeclaredFields()) {
+			patientFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+
+		Class person = Person.class;
+		Set<String> personFieldNames = new HashSet<String>(person.getDeclaredFields().length);
+		for (Field f : person.getDeclaredFields()) {
+			personFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+
+		Class personName = PersonName.class;
+
+		Set<String> personNameFieldNames = new HashSet<String>(personName.getDeclaredFields().length);
+		for (Field f : personName.getDeclaredFields()) {
+			personNameFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+
+		Class identifier = PatientIdentifier.class;
+		Set<String> identifierFieldNames = new HashSet<String>(identifier.getDeclaredFields().length);
+		for (Field f : identifier.getDeclaredFields()) {
+			identifierFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+		List<String> whereConditions = new ArrayList<String>();
+
+
+		List<String> innerFields = new ArrayList<String>();
+		String innerSelect = " from patient p1 ";
+
+		for (String s : attributes) {
+			if (s != null)
+				s = s.trim();
+			if (patientFieldNames.contains(s)) {
+
+				AbstractEntityPersister aep = ((AbstractEntityPersister) sessionFactory.getClassMetadata(Patient.class));
+				String[] properties = aep.getPropertyColumnNames(s);
+				if (properties.length >= 1)
+					s = properties[0];
+				whereConditions.add(" t1." + s + " = t5." + s);
+				innerFields.add("p1." + s);
+			} else if (personFieldNames.contains(s)) {
+				if (!outerSelect.contains("person")) {
+					outerSelect += "inner join person t2 on t1.patient_id = t2.person_id ";
```

Comment:
@dkayiwa and @wluyima I would be in favor of using hibernates criteria, since there is a lot of "manual" SQL construction going on here that is hard to read overall. With criteria we prevent some errors already at compile time and it makes it more readable in my view. Whats your opinion on this?